packages/next/build/webpack/plugins/next-esm-plugin.ts

```diff
@@ -0,0 +1,258 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2018 Prateek Bhatnagar
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * Webpack plugin for NextJs that runs a child compiler to generate a second (modern) JS bundle
+ *
+ * @author: Janicklas Ralph (https://github.com/janickals-ralph)
+ *
+ * Original source from which this was built upon - https://github.com/prateekbh/babel-esm-plugin
+ */
+import {
+  Compiler,
+  compilation,
+  Plugin,
+  RuleSetRule,
+  RuleSetLoader,
+  Output,
+} from 'webpack'
+
+const SingleEntryPlugin = require('webpack/lib/SingleEntryPlugin')
+const MultiEntryPlugin = require('webpack/lib/MultiEntryPlugin')
+const JsonpTemplatePlugin = require('webpack/lib/web/JsonpTemplatePlugin')
+const SplitChunksPlugin = require('webpack/lib/optimize/SplitChunksPlugin')
+const RuntimeChunkPlugin = require('webpack/lib/optimize/RuntimeChunkPlugin')
+
+const PLUGIN_NAME = 'NextEsmPlugin'
+const FILENAME = '[name].module.js'
+const CHUNK_FILENAME = '[id].module.js'
```

Comment:
Is this correct?

Our main webpack config uses the following names:
```js
      filename: ({ chunk }: { chunk: { name: string } }) => {
        // Use `[name]-[contenthash].js` in production
        if (
          !dev &&
          (chunk.name === CLIENT_STATIC_FILES_RUNTIME_MAIN ||
            chunk.name === CLIENT_STATIC_FILES_RUNTIME_WEBPACK)
        ) {
          return chunk.name.replace(/\.js$/, '-[contenthash].js')
        }
        return '[name]'
      },
      chunkFilename: isServer
        ? `${dev ? '[name]' : '[name].[contenthash]'}.js`
        : `static/chunks/${dev ? '[name]' : '[name].[contenthash]'}.js`,
```

Reply:
These are just placeholders. I initiate the plugin in a different way to handle Next's `filename` and `chunkFilename`

```
new NextEsmPlugin({
          filename: (getFileName: Function | string) => (...args: any[]) => {
            const name =
              typeof getFileName === 'function'
                ? getFileName(...args)
                : getFileName

            return name.includes('.js')
              ? name.replace(/\.js$/, '.module.js')
              : args[0].chunk.name.replace(/\.js$/, '.module.js')
          },
          chunkFilename: (inputChunkName: string) =>
            inputChunkName.replace(/\.js$/, '.module.js'),
        })
```